mypy/checker.py

```diff
@@ -1188,18 +1190,31 @@ def check_assignment(self, lvalue: Node, rvalue: Node, infer_lvalue_type: bool =
                         partial_types = self.find_partial_types(var)
                         if partial_types is not None:
                             if not self.current_node_deferred:
-                                var.type = rvalue_type
+                                if experiments.STRICT_OPTIONAL:
+                                    var.type = UnionType.make_simplified_union(
```

Comment:
The declared type of variable that gets assigned here is correct (`Union[t, None]` where `t` is the type of rvalue). However, the effective conditional type should be just `t` within the current block. For example, this should be valid:

```
x = None
if foo:
    x = 1
    print(x + 1)  # type of x is int here
# type of x is Optional[int] here
```